### PR TITLE
Implement ServiceModule (Part 2)

### PR DESCRIPTION
**Pull request by gshaw1997**
This PR is the **part 2** of a refactor to `library-service` which includes pulling download logic into it's own module.

A `ServiceModule` can be defined as a software component of a larger system that encapsulates a particular set of functionality. The `ServiceModule` is an abstract class with a re-usable set of functionality that a `ServiceModule` should have in order to be constructed and used. 

[Part 1 (DependencyContainer)](https://github.com/Cyber4All/library-service/pull/99)

### CODE REVIEW

File: src/shared/ServiceModule.ts [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+  private static providerKeys: InjectionKey[] = [];
+  private static _initialized: boolean = false;
+
+  protected static _adapter: any;
+
+  protected constructor() {}
+
+  /**
+   * Initializes module and it's dependencies
+   * Only one instance a ServiceModule can be initialized at a time.
```
Comment (DonSeannelly) [inline src/shared/ServiceModule.ts]:
```suggestion
   * Only one instance of a ServiceModule can be initialized at a time.
```

File: src/shared/ServiceModule.ts [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+   *
+   * @readonly
+   * @protected
+   * @static
+   * @type {boolean}
+   * @memberof ServiceModule
+   */
+  protected static get initialized(): boolean {
+    return this._initialized;
+  }
```
Comment (DonSeannelly) [inline src/shared/ServiceModule.ts]: When is it valuable for us to check if a module has been initialized? I think I understand the initialization/destroy process here, but maybe not the direct use cases.

File: src/shared/ServiceModule.ts [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+export function serviceModule({
+  adapter,
+  providers = []
+}: {
+  adapter: any;
+  providers: ClassProvider[];
+}) {
+  return function(target: any) {
+    target._adapter = adapter;
+    target.providers = providers;
```
Comment (DonSeannelly) [inline src/shared/ServiceModule.ts]: Why is adapter prefixed with an underscore, but providers is not?

Comment (gshaw1997) [reply src/shared/ServiceModule.ts]: `providers` can be publicly set and get while `adapter` can only be set by the class and get publicly.

File: src/shared/ServiceModule.ts (hunk 2 above)
Comment (gshaw1997) [reply src/shared/ServiceModule.ts]: Only the class and the sub-classes are able to use this method. The use case this is valuable in is to check if the module has been initialized, meaning it has all of its dependencies, before allowing access to the module's functionality via its adapters.

**Pull request merged**